### ISSUE TITLE
mpris-notifier doesn't emit notifications

### ISSUE THREAD

**Issue by killerbees19** [l1na-forever/mpris-notifier#1]
Tested with Firefox (YouTube Music) and VLC.

`notify-send` works and `dbus-monitor` shows the events.

Xubuntu 20.04, `mpris-notifier` built from source.

**Comment by l1na-forever** [comment]
Hey, thanks for the report! I've repro'd both bugs.

In both cases, notifications aren't being emitted for two reasons:

- `PropertiesChanged` should fire for either `PlaybackStatus` or `Metadata` changing. The clients I was testing with coalesced these into a single DBus message, but it looks like VLC and FF send them separately. Will fix this assumption.
- Both VLC and FF emit unhelpful values for `trackId`. It [should be unique](https://specifications.freedesktop.org/mpris-spec/latest/Player_Interface.html#Simple-Type:Track_Id), but FF only emits a constant ID. I'll work around this with a fix. 

I'll let you know when the changes are ready, sorry for the trouble.

**Comment by l1na-forever** [comment]
TODO:

- [x] accept `PropertiesChanged` signals with either/both `PlaybackStatus`/`Metadata`
- [x] get rid of the `trackId` check
- [x] support `artUrl` props with `file://` URLs, as firefox sends
- [x] drop notifications that would emit an empty string, depending on the format string
- [x] cache metadata/playerstatus across senders

**Comment by l1na-forever** [comment]
@killerbees19 Hey, I [just released `v0.1.3`](https://github.com/l1na-forever/mpris-notifier/releases/tag/v0.1.3)! If you do a `cargo install --force mpris-notifier`, you should get the latest version. Please let me know if it doesn't resolve resolve!

**Comment by killerbees19** [comment]
Thanks! :-)

btw: `version` in `Cargo.toml` is still at `0.1.2` at `mainline` branch. (Latest commit: 26e8c7dee0f8e5d4a9061c5278443e35f1465e7f)

* Chromium: A real mess with too many events/notifications. (Probably not your fault!)
* VLC: Play/Pause works now, but sadly **next/prev track event is still ignored**.
  And sometimes even play/pause is missing, but I can't reproduce the last issue.
  Tested with four local MP3 files.
* Firefox: Seems to work, but it's a little bit annoying with multiple notifications for a single track change.
  Tested with YouTube Music, [joshwoodward.com](https://www.joshwoodward.com/) and some other sites.
  I don't know if that's something you could work around in your code…

![Bildschirmfoto_2022-07-27_14-55-53](https://user-images.githubusercontent.com/21372289/181253566-a17b4fc0-d567-40cc-8b8a-48e50bb4d70f.png)

![Bildschirmfoto_2022-07-27_14-40-57](https://user-images.githubusercontent.com/21372289/181253590-40e50ee1-d522-4879-8c12-6720b878920b.png)

![Bildschirmfoto_2022-07-27_14-41-17](https://user-images.githubusercontent.com/21372289/181253615-967e0e03-109e-43d1-a3ec-6fdd7ede4be7.png)

Each screenshot shows a single track change with no old notifications visible:

1. The old (previous) track again?
2. An incomplete notification without album and cover for the next (current) track.
3. A complete notification for the currently played track.
4. Another incomplete notification without album.

Notifications in screenshots: From bottom (1) to top (4)

Anyways, I think the browser "bugs" are not important and just cosmetic glitches.

**Comment by l1na-forever** [comment]
Thanks for your persistence in testing this, it's really appreciated! :) 

> btw: version in Cargo.toml is still at 0.1.2 at mainline branch. (Latest commit: https://github.com/l1na-forever/mpris-notifier/commit/26e8c7dee0f8e5d4a9061c5278443e35f1465e7f)

Oops, thanks for the heads up! Pushed the last bits here.

> Chromium: A real mess with too many events/notifications. (Probably not your fault!)

Will take a look into Chromium, it might just be that I have to figure out a slick debounce scheme to make it work.

> VLC: Play/Pause works now, but sadly next/prev track event is still ignored.

Took a look and oof, found out why. When I do prev/next in VLC, it emits a Stopped event, without a new Playing event for the track starting:

```
signal time=1658946826.852412 sender=:1.42 -> destination=(null destination) serial=36 path=/org/mpris/MediaPlayer2; interface=org.freedesktop.DBus.Properties; member=PropertiesChanged
   string "org.mpris.MediaPlayer2.Player"
   array [      
      dict entry(         
         string "PlaybackStatus"               
         variant             string "Stopped"               
      )         
   ]                      
   array [                              
   ]                                                  
signal time=1658946827.050926 sender=:1.42 -> destination=(null destination) serial=37 path=/org/mpris/MediaPlayer2; interface=org.freedesktop.DBus.Properties; member=PropertiesChanged
   string "org.mpris.MediaPlayer2.Player"
   array [      
      dict entry(         
         string "Metadata"           
         variant             array [                       
               dict entry(
                  string "mpris:trackid"
                  variant                      object path "/org/videolan/vlc/playlist/3"
               )                                      
               dict entry(

...
```

I don't think I can key off the `Stopped` signal reliably. I'll change the logic so that a new track change is always notified, whereas a 'playing' notification will check to make sure the notification hasn't popped up before.

> Firefox: Seems to work, but it's a little bit annoying with multiple notifications for a single track change.

Ah, let me test with that site as well, I was testing with the Spotify web console. I'll take another stab at getting rid of duplicate notifications today and see if I can get browser output cleaned up :sparkles:

**Comment by l1na-forever** [comment]
Yeah, it looks like what's going on is that (sometimes!) some players emit multiple PropertiesChanged notifications for one track change. Here's Firefox filling in the album field after sending the first:

```
signal time=1658956366.404775 sender=:1.31 -> destination=(null destination) serial=99 path=/org/mpris/MediaPlayer2; interface=org.freedesktop.DBus.Properties; member=PropertiesChanged
   string "org.mpris.MediaPlayer2.Player"
   array [
      dict entry(
         string "Metadata"
         variant             array [
               dict entry(
                  string "mpris:trackid"
                  variant                      object path "/org/mpris/MediaPlayer2/firefox"
               )
               dict entry(
                  string "xesam:title"
                  variant                      string "Tears Don't Fall • Bullet For My Valentine"
               )
               dict entry(
                  string "xesam:album"
                  variant                      string ""
               )
               dict entry(
                  string "xesam:artist"
                  variant                      array [
                        string ""
                     ]
               )
            ]
      )
   ]
   array [
   ]
signal time=1658956366.410412 sender=:1.31 -> destination=(null destination) serial=100 path=/org/mpris/MediaPlayer2; interface=org.freedesktop.DBus.Properties; member=PropertiesChanged
   string "org.mpris.MediaPlayer2.Player"
   array [
      dict entry(
         string "Metadata"
         variant             array [
               dict entry(
                  string "mpris:trackid"
                  variant                      object path "/org/mpris/MediaPlayer2/firefox"
               )
               dict entry(
                  string "xesam:title"
                  variant                      string "Tears Don't Fall"
               )
               dict entry(
                  string "xesam:album"
                  variant                      string "The Poison"
               )
               dict entry(
                  string "xesam:artist"
                  variant                      array [
                        string "Bullet For My Valentine"
                     ]
               )
            ]
      )
   ]
   array [
   ]
```

I have a slightly different design in mind to work around this which should fix the "notifications missing half their stuff" bug.

**Comment by l1na-forever** [comment]
Update: newer design seems to work much more smoothly with the way browsers send their notifications, I've been testing it with Spotify/YouTube/www.joshwoodward.com and I haven't been able to trigger duplicate or half-filled notifications. Cleaning up first, but change should be out later.

**Comment by l1na-forever** [comment]
@killerbees19  hey, I just released `v0.1.4` to Cargo, which should hopefully fix Firefox/VLC at least (I'm optimistic that it might fix other chatty clients as well). I will say I've noticed a lot of strange behavior from Firefox's MPRIS bits, where sometimes it won't emit any track metadata for plain old music embeds (like on joshwoodward.com), but YouTube seems smooth. Thanks again for your patience!

**Comment by killerbees19** [comment]
Nice, looks much better! 👍

Sometimes there are still additional notifications in the middle of a song or when the next track starts. But it's possible that especially YouTube (Video & Music) has some strange tricks in place that distract the browser. Or the MPRIS implementation of Firefox sucks. I don't know…

Anyways, VLC looks fine so far. I'll test it for some days and report back if there are other issues left. I'll close this issue now, because the initial bug is fixed.

Thank you for this great piece of software ☺
